src/coreclr/jit/emitarm64.cpp

```diff
@@ -16196,55 +16324,154 @@ bool emitter::IsRedundantLdStr(
+                //    - lvaVarNum = current instruction's lclVar
+                //    - lvaVarNum2 = last instruction's lclVar
+                //    - lvaOffset = current instruction's offset
+
+                if (optimizationOrder == eRO_ascending)
+                {
+                    emitIns_SS_R_R_R_I(optIns, prevReg1Attr, reg1Attr, prevReg1, reg1, reg2, prevImmSize, prevLclVarNum,
+                                       prevOffset, varx);
+                }
+                else
+                {
+                    emitIns_SS_R_R_R_I(optIns, reg1Attr, prevReg1Attr, reg1, prevReg1, reg2, newImmSize, varx, offs,
+                                       prevLclVarNum);
+                }
+                return true;
+            }
+            else if (isLastGCLclVar)
+            {
+                // if (ascending)
+                //    - idGCref() = last instruction's GCtype
+                //    - idGCrefReg2() = GCT_NONE
+                //    - lvaVarNum = last instruction's lclVar
+                //    - lvaOffset = last instruction's offset
+                // else
+                //    - idGCref() = GCT_NONE
+                //    - idGCrefReg2() = last instruction's GCtype
+                //    - lvaVarNum = last instruction's lclVar
+                //    - lvaOffset = current instruction's offset
+
+                if (optimizationOrder == eRO_ascending)
+                {
+                    emitIns_SS_R_R_R_I(optIns, prevReg1Attr, reg1Attr, prevReg1, reg1, reg2, prevImmSize, prevLclVarNum,
+                                       prevOffset);
+                }
+                else
+                {
+                    emitIns_SS_R_R_R_I(optIns, reg1Attr, prevReg1Attr, reg1, prevReg1, reg2, newImmSize, prevLclVarNum,
+                                       prevOffset);
+                }
+                return true;
+            }
+            else if (isCurrGCLclVar)
+            {
+                // if (ascending)
+                //    - idGCref() = GCT_NONE
+                //    - idGCrefReg2() = current instruction's GCType
+                //    - lvaVarNum = current instruction's lclVar
+                //    - lvaOffset = last instruction's offset
+                // else
+                //    - idGCref() = current instruction's GCType
+                //    - idGCrefReg2() = GCT_NONE
+                //    - lvaVarNum = current instruction's lclVar
+                //    - lvaOffset = current instruction's offset
+
+                if (optimizationOrder == eRO_ascending)
+                {
+                    emitIns_SS_R_R_R_I(optIns, prevReg1Attr, reg1Attr, prevReg1, reg1, reg2, prevImmSize, varx, offs);
+                }
+                else
+                {
+                    emitIns_SS_R_R_R_I(optIns, reg1Attr, prevReg1Attr, reg1, prevReg1, reg2, newImmSize, varx, offs);
+                }
+                return true;
+            }
```

Comment:
Two things:
1. I don't think you need these three cases. If either of the vars is a lclvar, then use the same code path and take the appropriate ascending/descending case, passing the var1/var2 offs1/offs2 as appropriate.
2. I don't think you should distinguish between GC vars and non-GC vars. Currently, non-GC vars go through the R_R_R_I case, which is why we lose var info in the instruction (that could be used in disasm output). (maybe you'd need to implement the `emitGCvarDeadUpd()` code path support in `emitOutputInstr` for LclVarPairs?)

Reply:
> I don't think you should distinguish between GC vars and non-GC vars

Ah, you are suggesting to use `emitIns_SS_R_R_I()` method even for non-GC vars `stp`? Yes, that makes sense. We could also use it for `ldp` which will fix https://github.com/dotnet/runtime/issues/84504.

> maybe you'd need to implement the `emitGCvarDeadUpd()` code path support in emitOutputInstr for LclVarPairs?

I am not sure I follow this. Are you suggesting having a newly written version of `emitGCvarDeadUpd` for lclVarPairs that will update both variables in the same method, instead of what we do currently, where we update them in two separate calls of `emitGCvarDeadUpd()`?

Reply:
fixed.